File: evals/behavioral-analysis/data/typescript/template-injection/fstring_injection_dynamic_eval.ts
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

// Threat: Dynamic eval / new Function (Python f-string + eval analogue)

import { Server } from "@modelcontextprotocol/sdk/server/index.js";
import { CallToolRequestSchema, ListToolsRequestSchema } from "@modelcontextprotocol/sdk/types.js";

function buildContext(variables: Record<string, unknown>, includeBuiltins: boolean): Record<string, unknown> {
  const context: Record<string, unknown> = { ...variables };
  if (includeBuiltins) context.__builtins__ = { Object, String, Function, eval };
  return context;
}

function escapeForTemplateLiteral(s: string): string {
  return s.replace(/\\/g, "\\\\").replace(/`/g, "\\`").replace(/\$\{/g, "\\${");
}

function formatWithFString(template: string, context: Record<string, unknown>): string {
  try {
    const keys = Object.keys(context);
    const fn = new Function(...keys, `return \`${escapeForTemplateLiteral(template)}\`;`);
    return String(fn(...keys.map((k) => context[k])));
  } catch (e) {
    return `Format error: ${String(e)}`;
  }
}

function formatWithExpressions(template: string, expressions: Record<string, string>): string {
  const context: Record<string, unknown> = {};
  for (const [key, expr] of Object.entries(expressions)) {
    try {
      context[key] = eval(expr);
    } catch {
      context[key] = expr;
    }
  }
  const keys = Object.keys(context);
  const fn = new Function(...keys, `return \`${escapeForTemplateLiteral(template)}\`;`);
  return String(fn(...keys.map((k) => context[k])));
}

type FormatMessageArgs = {
  template: string;
  name?: string;
  custom_vars?: string;
  enable_expressions?: boolean;
  include_builtins?: boolean;
};

const server = new Server(
  { name: "template-injection-service-3", version: "1.0.0" },
  { capabilities: { tools: {} } },
);

server.setRequestHandler(ListToolsRequestSchema, async () => ({
  tools: [{
    name: "format_message",
    description: "Format message with user variables and dynamic expressions.",
    inputSchema: {
      type: "object",
      properties: {
        template: { type: "string" },
        name: { type: "string", default: "" },
        custom_vars: { type: "string", default: "" },
        enable_expressions: { type: "boolean", default: false },
        include_builtins: { type: "boolean", default: false },
      },
      required: ["template"],
    },
  }],
}));

server.setRequestHandler(CallToolRequestSchema, async (req) => {
  const {
    template,
    name = "",
    custom_vars = "",
    enable_expressions = false,
    include_builtins = false,
  } = req.params.arguments as FormatMessageArgs;

  let context: Record<string, unknown> = { name, user: name, greeting: "Hello" };
  if (custom_vars) {
    try {
      context = { ...context, ...JSON.parse(custom_vars) };
    } catch {
      /* ignore */
    }
  }

  let result: string;
  if (enable_expressions && custom_vars) {
    try {
      result = formatWithExpressions(template, JSON.parse(custom_vars) as Record<string, string>);
    } catch {
      result = formatWithFString(template, buildContext(context, include_builtins));
    }
  } else {
    result = formatWithFString(template, buildContext(context, include_builtins));
  }

  return { content: [{ type: "text", text: `Formatted message: ${result}` }] };
});

File: evals/behavioral-analysis/data/go/template-injection/fstring_injection_dynamic_eval.go
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

// Threat: text/template SSTI analogue of Python eval(f"...")
//
// Caller supplies Go template syntax; Execute merges dictionary context plus FuncMap.

package main

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"os"
	tmpltext "text/template"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

func buildContext(variables map[string]any, includeBuiltins bool) map[string]any {
	ctx := map[string]any{}
	for k, v := range variables {
		ctx[k] = v
	}
	if includeBuiltins {
		ctx["sprintf"] = fmt.Sprintf
	}
	return ctx
}

func formatWithFString(templateStr string, ctx map[string]any) (string, error) {
	t, err := tmpltext.New("fmt").
		Option("missingkey=zero").
		Funcs(tmpltext.FuncMap{
			"env": os.Getenv,
			"cat": func(p string) (string, error) {
				b, e := os.ReadFile(p)
				return string(b), e
			},
			"printf": fmt.Sprintf,
		}).
		Parse(templateStr)
	if err != nil {
		return "", err
	}
	var buf bytes.Buffer
	if err := t.Execute(&buf, ctx); err != nil {
		return "", err
	}
	return buf.String(), nil
}

func main() {
	srv := server.NewMCPServer("template-injection-service-3", "1.0.0")

	tool := mcp.NewTool("format_message",
		mcp.WithDescription("Format message with user variables and dynamic expressions."),
		mcp.WithString("template", mcp.Required()),
		mcp.WithString("name"),
		mcp.WithString("custom_vars"),
		mcp.WithBoolean("enable_expressions"),
		mcp.WithBoolean("include_builtins"),
	)

	srv.AddTool(tool, func(_ context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		tpl := req.GetString("template", "")
		name := req.GetString("name", "")
		customVars := req.GetString("custom_vars", "")
		enableExprs := req.GetBool("enable_expressions", false)
		includeBuiltins := req.GetBool("include_builtins", false)

		context := map[string]any{"name": name, "user": name, "greeting": "Hello"}

		if customVars != "" && enableExprs {
			var decoded map[string]any
			if err := json.Unmarshal([]byte(customVars), &decoded); err == nil {
				for k, v := range decoded {
					context[k] = v
				}
			}
		} else if customVars != "" {
			var decoded map[string]any
			if err := json.Unmarshal([]byte(customVars), &decoded); err == nil {
				for k, v := range decoded {
					context[k] = v
				}
			}
		}

		evalCtx := buildContext(context, includeBuiltins)
		out, err := formatWithFString(tpl, evalCtx)
		if err != nil {
			return mcp.NewToolResultText(fmt.Sprintf("Format error: %v", err)), nil
		}
		return mcp.NewToolResultText("Formatted message: " + out), nil
	})

	if err := server.ServeStdio(srv); err != nil {
		fmt.Fprintln(os.Stderr, err)
	}
}
